rollup-interface/src/state_machine/da.rs

```diff
@@ -56,14 +58,68 @@ pub trait DaVerifier {
     ) -> Result<(), Self::Error>;
 }
 
+// Simple structure that implements the Read trait for a buffer and  counts the number of bytes read from the beginning
+// Useful for the partial blob reading optimization: we know for each blob how many bytes have been read from the beginning
+// Because of soundness issues we cannot implement the Buf trait because the prover could get unproved blob data using the chunk method.
+#[derive(Debug, Clone, Serialize, Deserialize, BorshDeserialize, BorshSerialize, PartialEq)]
+pub struct BufReaderWithCounter<B: Buf> {
```

Comment:
nit: Since this type shows up in the public API, let's make sure that it's as clean and well-documented as we can. 

I'd recommend adding a doc comment to each field, and maybe renaming to `CountedBufReader`